tests/models/test_genericconfig.py

```diff
@@ -35,36 +62,92 @@ class SomeSubclass(GenericConfig):
+    """Ensure KeyError is raised for a non-existent key."""
+    non_existent_key = "non_existent_key"
+    assert non_existent_key not in generic_config.configuration
+    with pytest.raises(KeyError):
+        generic_config.configuration[non_existent_key]
+
+
+def test_attribute_set(generic_config: "CustomConfig") -> None:
+    """Test configuration.__setitem__, and thus also configuration.__setattr__.
+
+    Assign different values to test dynamic type-casting.
+    If a field has been specified to have a specific type, the incoming value should be
+    cast to this type. However, if it is a non-specified type (similar to a standard
+    dict key/value-pair), then it should be fine to change the value type for that
+    given key.
 
-    Assign both a valid and an invalid value, to test dynamic
-    type-checking.
     """
     generic_config.configuration["string"] = "bar"
-    try:
-        generic_config.configuration["string"] = 3.14
-        assert False
-    except TypeError:
-        pass
+    generic_config.configuration["string"] = 3.14
+    assert generic_config.configuration["string"] == str(3.14)
+
+    generic_config.configuration["float"] = "0"
+    assert generic_config.configuration["float"] == "0"
 
 
-def test_attribute_contains(generic_config: "GenericConfig") -> None:
+def test_attribute_contains(generic_config: "CustomConfig") -> None:
     """Test confguration.__contains__."""
     assert "float" in generic_config.configuration
 
 
-def test_attribute_del_item(generic_config: "GenericConfig") -> None:
+def test_attribute_del_item(generic_config: "CustomConfig") -> None:
     """Test configuration.__delitem__."""
     del generic_config.configuration["float"]
     assert "float" not in generic_config.configuration
 
+    # "string" is defined as a pydantic model field.
+    # This means it should not be deleted, but reset to its default value
+    # and removed from the set of set fields.
+    generic_config.configuration["string"] = "my secret string"
+    assert generic_config.configuration.string == "my secret string"
+    assert "string" in generic_config.configuration.__fields_set__
+
+    del generic_config.configuration["string"]
+    assert "string" in generic_config.configuration
+    assert (
+        generic_config.configuration.string
+        == generic_config.configuration.__fields__["string"].default
+    )
+    assert "string" not in generic_config.configuration.__fields_set__
+
+
+def test_attribute_del_item_fail(generic_config: "CustomConfig") -> None:
+    """Ensure KeyError is raised if key does not exist in AttrDict."""
+    non_existent_key = "non_existant_key"
+    assert non_existent_key not in generic_config.configuration
+    with pytest.raises(KeyError):
+        del generic_config.configuration[non_existent_key]
+
+    # Make sure deleting a pydantic field does not fully delete it
+    # and hence does not raise KeyError.
```

Comment:
Why should this not be fully deleted?

Reply:
Yeah that's maybe not clear.
But it should not be fully deleted, because deleting it completely removes the notion of that pydantic attribute/field in the model - that's not something we want. We want the pydantic model to stay intact - instead when deleting a pydantic field, we should reset it to the value it would've had if it was not supplied originally.
At least that's my take on it.